src/commands/branch.rs

```diff
@@ -294,6 +301,40 @@ fn cmd_branch_list(
         }
     }
 
+    if !args.revisions.is_empty() {
+        fn both_targets(branch_target: &BranchTarget) -> impl Iterator<Item = CommitId> + '_ {
+            branch_target
+                .local_target
+                .iter()
+                .chain(branch_target.remote_targets.values())
+                .flat_map(move |target| target.adds())
+        }
+
+        // Intersects with the set of all branch targets to minimize the lookup space.
+        // It also ensures that hidden targets are populated.
+        let filter_expressions: Vec<_> = args
+            .revisions
+            .iter()
+            .map(|revision_str| workspace_command.parse_revset(revision_str))
+            .try_collect()?;
+        let filter_expression = RevsetExpression::union_all(&filter_expressions);
+        let all_targets = all_branches.values().flat_map(both_targets).collect();
+        let revset_expression =
+            RevsetExpression::commits(all_targets).intersection(&filter_expression);
+        // Optimize out implicit "all() &" from negative filter expression.
+        let revset_expression = revset::optimize(revset_expression);
+        let revset = workspace_command.evaluate_revset(revset_expression)?;
+        let filtered_targets: HashSet<CommitId> = revset.iter().collect();
+        // TODO: If we add name-based filter like --glob, this might have to be
+        // rewritten as a positive list or predicate function. Should they
+        // be AND-ed or OR-ed? Maybe OR as "jj git push" would do. Perhaps, we
```

Comment:
I think they should definitely be ORed, just like repeated `-r`-s.

OTOH, if the `branches` revset doesn't yet take globs, it could, so that you could use `-r` to AND if you really wanted to.

Reply:
Let's delete the "should they be AND or ORed" part of the comment? Instead, we could say "these should be ORed, just like repeated `-r`-s".

Reply:
I don't think it's 100% clear. If you consider that `-r` is different axis than branch names (something like `jj log -r` and `jj log PATH`), these two predicates should be intersected.

So let the future implementer decide.